File: 0722-remove-comments/0722-remove-comments.py
class Solution:
    def removeComments(self, source: List[str]) -> List[str]:        
        # object to be returned
        out_put = []
        
        # create a flag to know if in comment
        in_comment = False
        
        # iterate the source code
        for line in source:
            if not in_comment:
                out_put.append([])
            i = 0
            while i < len(line):
                char = line[i]
                if in_comment:
                    if char == '*' and i + 1 < len(line) and line[i + 1] == '/':
                        in_comment = False
                        i += 1
                    i += 1
                else:
                    if char == '/' and i + 1 < len(line) and line[i + 1] == '*':
                        in_comment = True
                        i += 1
                    elif char == '/' and i + 1 < len(line) and line[i + 1] == '/':
                        break
                    else:
                        out_put[-1].append(char)
                    i += 1
            if not in_comment:
                new_line = "".join(out_put.pop())
                if new_line:
                    out_put.append(new_line)
                    
        return out_put
                    

File: 0722-remove-comments/0722-remove-comments.js
/**
 * @param {string[]} source
 * @return {string[]}
 */
var removeComments = function removeComments(source) {
  // object to be returned
  let out_put = [];

  // create a flag to know if in comment
  let in_comment = false;

  // iterate the source code
  for (let line of source) {
    if (!in_comment) {
      out_put.push([]);
    }
    let i = 0;
    while (i < line.length) {
      let char = line.charAt(i);
      if (in_comment) {
        if (char === '*' && i + 1 < line.length && line.charAt(i + 1) === '/') {
          in_comment = false;
          i += 1;
        }
        i += 1;
      } else {
        if (char === '/' && i + 1 < line.length && line.charAt(i + 1) === '*') {
          in_comment = true;
          i += 1;
        } else if (char === '/' && i + 1 < line.length && line.charAt(i + 1) === '/') {
          break;
        } else {
          out_put[out_put.length - 1].push(char);
        }
        i += 1;
      }
    }
    if (!in_comment) {
      let new_line = out_put.pop().join('');
      if (new_line) {
        out_put.push(new_line);
      }
    }
  }
  return out_put;
}